### PR TITLE
nixos/tayga: add mappings option

### LINKED ISSUE #302381: Provided tayga service doesn't support SIIT-DC operation (with explicit mappings)
### Describe the bug

When using the provided tayga service, it's not really possible to setup explicit mappings. This seems like an oversight.

### Steps To Reproduce
N/A

### Expected behavior
Some sort of `services.tayga.mappings` should exist.

### Screenshots
N/A

### Additional context
It is possible to workaround the issue by sticking them into

```
services.tayga.ipv4.address =
  ''
    192.0.2.1
    map 198.51.100.1 2001:db8::1
  '';
```

but... uh yeah anyway. :sweat_smile: 

### Notify maintainers

@hax404 @livingshredder (?)

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 6.1.72, NixOS, 23.11 (Tapir), 23.11.2936.428544ae95ee`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.11"`
 - channels(nixos): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by hax404**
## Description of changes
Adds mappings option for static mappings between IPv4 and IPv6.

fixes https://github.com/NixOS/nixpkgs/issues/302381

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/tayga.nix [hunk 1]
```diff
@@ -103,13 +105,20 @@ in
...
 
       tunDevice = mkOption {
         type = types.str;
         default = "nat64";
-        description = "Name of the nat64 tun device";
+        description = "Name of the nat64 tun device.";
+      };
+
+      mappings = mkOption {
+        type = types.listOf types.str;
```
Comment (hexchen) [inline nixos/modules/services/networking/tayga.nix:118]: Would it potentially make more sense for the type to be an `attrsOf str`? Or is it possible to have multiple IPv6 addresses mapped to the same IPv4 address?

File: nixos/modules/services/networking/tayga.nix [hunk 2]
```diff
@@ -103,13 +105,20 @@ in
       dataDir = mkOption {
         type = types.path;
         default = "/var/lib/tayga";
-        description = "Directory for persistent data";
+        description = "Directory for persistent data.";
```
Comment (hexchen) [inline nixos/modules/services/networking/tayga.nix:108]: I'm unsure about the trailing dots, any specific reason you added them?

**Review by hexchen — COMMENTED** [review]
Seems reasonable to me, only have a few minor requests/questions that shouldn't be interpreted as blockers.

File: nixos/modules/services/networking/tayga.nix (hunk 2 above)
Comment (hax404) [reply nixos/modules/services/networking/tayga.nix:108]: Just to make it consistent in this module.

File: nixos/modules/services/networking/tayga.nix (hunk 1 above)
Comment (hax404) [reply nixos/modules/services/networking/tayga.nix]: Neither IPv4 nor IPv6 addresses are accepted twice in the mapping by tayga. So I've changed the module and added a check.

**Review by hexchen — APPROVED** [review]
LGTM

**Pull request merged**